### ISSUE TITLE
ch4: Initialization problems on platforms with more than 8 NICs

### ISSUE THREAD

**Issue by wenduwan** [pmodels/mpich#6688]
#### 1. NIC limit
ch4 currently has a hardcoded [8 NICs limit](https://github.com/pmodels/mpich/blob/16e1e6002d401a43edaba10c9b8bf3ee83824a5c/src/mpid/ch4/netmod/ofi/ofi_pre.h#L287) that cannot be changed at runtime. As a result, on platforms with 8+ NICs the user cannot fully utilize the available bandwidth. To make their lives harder, there is a also a bug that causes segfault that prevents `MPI_Init` entirely(will post PR shortly).

#### 2. Address exchange
In addition, we have also noticed a high volume of `fi_av_insert` calls [as part of ](https://github.com/pmodels/mpich/blob/16e1e6002d401a43edaba10c9b8bf3ee83824a5c/src/mpid/ch4/netmod/ofi/ofi_init.c#L759)`MPI_Init`, where each rank inserts `O(comm_size x num_nic x num_vci)` addresses to the address vector. This is problematic for libfabric providers with low AV size limits, e.g. EFA currently only supports max 256 local peers(which translates to `-ppn 8` x 32 nics, i.e. some nics cannot be used). 

### The Ask

Regarding 1), we want to explore the possibility to eliminate the hardcoded limit, e.g. Open MPI does not have such. We also understand that this will likely take some time, so we propose that for the short term we should raise the limit to e.g. 128, if that's not harmful.

On 2), first we want to understand the purpose of the current address exchange process, and then improve it to minimize the number of addresses.

### Environment
- Hardware: AWS p5.48xlarge instance (32 x EFA NICs)
- Operating System: Amazon Linux 2
- libfabric version: https://github.com/ofiwg/libfabric/commit/5342b3db0e36b04df91a3bb4dd3b1c35852b3b5c

**Comment by hzhou** [comment]
> On 2), first we want to understand the purpose of the current address exchange process, and then improve it to minimize the number of addresses.

It moves the overhead of av insertion to init time and prevents performance issues at runtime. It also significantly simplifies the code. Lazy address exchange will be quite complicated and potentially trigger more bugs and performance issues, but it may be necessary for a more general user base such as one for the cloud environment. We'll look into it.

**Comment by hzhou** [comment]
> Hardware: AWS p5.48xlarge instance (32 x EFA NICs)

It is hard to imagine that one process wants to use 32 NICs concurrently. If  `ppn > 1`, does it make sense to distribute the NICs to each process? I am not sure if this is supported currently, but it is easy to support.

**Comment by shijin-aws** [comment]
@hzhou Hello, it seems I am not seeing the segmentation fault @wenduwan was hitting on the 32 NIC platform with MPICH 5.0.1, but still hit it with MPICH 4.3.2 (the latest MPICH 4.x). Would u mind sharing what change between MPICH 4 and 5 fixed it? 


Side note: With MPICH 5 it seems it's not utilizing all 32 NICs's bandwidth still when I have 96 ranks per node, but at least it doesn't crash

**Comment by hzhou** [comment]
It might be https://github.com/pmodels/mpich/pull/7730 and https://github.com/pmodels/mpich/pull/7815. We skipped the multinic in #7730, which is probably the reason that it avoided crash but only used a single nic. #7815 fixed the root issue that we were not recognizing `HWLOC_OBJ_DIE`.

**Comment by shijin-aws** [comment]
Thanks @hzhou !

>it avoided crash but only used a single nic.

You mean each rank will use a single NIC right? as https://github.com/pmodels/mpich/pull/7730 says

>In the multinic case, provider may provide an invalid "null" pci info, which will result in hwloc failing to obtain topology. Rather than dealing this invalid case in the topology code, let's guard this case and deal with it in the higher layer. In the case of ofi multi-nic, we will simply treat all nics are equally close and equally distribute them among the ranks.

So it's like a round-robin assignment of NICs, if you have 32 NICs while 96 ranks, finally it's like 3 ranks use the same NIC?

**Comment by hzhou** [comment]
@shijin-aws You are right. It is supposed to be a round-robin assignment of NICs for each rank. Could you post the output of a dummy  mpi program with `MPIR_CVAR_DEBUG_SUMMARY=1`? That should dump the provider selection info during MPI_Init and may give a clue on what's going on.
